src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -129,58 +135,87 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
+            tasks.Add(SynchronizeProjectAssetAsync<CompilationOptions>(static p => p.CompilationOptions, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetAsync<ParseOptions>(static p => p.ParseOptions, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<ProjectReference>(static p => p.ProjectReferences, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<MetadataReference>(static p => p.MetadataReferences, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<AnalyzerReference>(static p => p.AnalyzerReferences, cancellationToken));
+
+            // Then sync each project's documents in parallel with each other.
+            foreach (var projectChecksums in allProjectChecksums)
+                tasks.Add(SynchronizeProjectDocumentsAsync(projectChecksums, cancellationToken));
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
+        }
+
+        static void AddAll(HashSet<Checksum> checksums, ChecksumCollection checksumCollection)
+        {
+            foreach (var checksum in checksumCollection)
+                checksums.Add(checksum);
+        }
+
+        async Task SynchronizeProjectAssetAsync<TAsset>(Func<ProjectStateChecksums, Checksum> getChecksum, CancellationToken cancellationToken)
+        {
+            await Task.Yield();
             using var _ = PooledHashSet<Checksum>.GetInstance(out var checksums);
 
-            checksums.Add(projectChecksums.Info);
-            checksums.Add(projectChecksums.CompilationOptions);
-            checksums.Add(projectChecksums.ParseOptions);
-            AddAll(checksums, projectChecksums.ProjectReferences);
-            AddAll(checksums, projectChecksums.MetadataReferences);
-            AddAll(checksums, projectChecksums.AnalyzerReferences);
-            AddAll(checksums, projectChecksums.Documents.Checksums);
-            AddAll(checksums, projectChecksums.AdditionalDocuments.Checksums);
-            AddAll(checksums, projectChecksums.AnalyzerConfigDocuments.Checksums);
+            foreach (var projectChecksums in allProjectChecksums)
+                checksums.Add(getChecksum(projectChecksums));
 
-            // First synchronize all the top-level info about this project.
-            await this.SynchronizeAssetsAsync<object, VoidResult>(
-                assetPath: AssetPath.ProjectAndDocuments(projectChecksums.ProjectId), checksums, callback: null, arg: default, cancellationToken).ConfigureAwait(false);
+            await this.SynchronizeAssetsAsync<TAsset, VoidResult>(
+                AssetPath.SolutionAndTopLevelProjectsOnly, checksums, callback: null, arg: default, cancellationToken).ConfigureAwait(false);
+        }
 
-            checksums.Clear();
+        async Task SynchronizeProjectAssetCollectionAsync<TAsset>(Func<ProjectStateChecksums, ChecksumCollection> getChecksums, CancellationToken cancellationToken)
```

Comment:
this is almost the same as the prior method.  just the primary method says "we have one checksum of a particular type per project" whereas thsi method says "we have many checksums of a particular type per project". 

Note; lots of checksums will be duplicated across projects (like metadata and analyzer references) so thsi hashset dedupes a lot most of the time.